src/data/matrix/notation.lean

```diff
@@ -39,13 +47,97 @@ variables {α : Type u} {o n m : ℕ} {m' n' o' : Type*}
 
 open_locale matrix
 
+/-- Matrices can be reflected whenever their entries can. We insert an `@id (matrix m' n' α)` to
+prevent immediate decay to a function. -/
+meta instance matrix.reflect [reflected_univ.{u}] [reflected_univ.{u_1}] [reflected_univ.{u_2}]
+  [reflected _ α] [reflected _ m'] [reflected _ n']
+  [h : has_reflect (m' → n' → α)] : has_reflect (matrix m' n' α) :=
+λ m, (by reflect_name : reflected _ @id.{(max u_1 u_2 u) + 1}).subst₂
+  ((by reflect_name : reflected _ @matrix.{u_1 u_2 u}).subst₃ `(_) `(_) `(_)) $
+  by {dunfold matrix, exact h m}
+
+section parser
+open lean
+open lean.parser
+open interactive
+open interactive.types
+
+/-- A version of `sep_by` that allows trailing delimiters, but requires at least one item -/
+private meta def sep_by' {α : Type} : parser unit → parser α → parser (list α)
+| s p := do
+  fst ← p,
+  some () ← optional s | pure [fst],
+  some rest ← optional (sep_by' s p) | pure [fst],
+  pure (fst :: rest)
+
+/-- A version of `many` that requires at least one -/
+private meta def at_least_one {α : Type} (p : parser α) : parser (list α) :=
+list.cons <$> p <*> (many p)
```

Comment:
I'd like to see these in a higher-up file, maybe `tactic/core.lean`.

Reply:
Done!